python/lsst/ci/prepare.py

```diff
@@ -384,6 +384,7 @@ def _fetch(self, product):
             #     git.fetch("origin", "--force", "--tags")
             # but avoids the overhead of two (possibly remote) git calls.
             git.fetch("-fup", "origin", "+refs/heads/*:refs/heads/*", "refs/tags/*:refs/tags/*")
+            git.fetch("-p")
```

Comment:
This is scary git voodoo that leaves the git repo with a broken `HEAD`.  This should have a detailed explanation of the problem it is solving and how it works.

Reply:
Given that the comment above says that we are trying to avoid doing two remote git calls and now we are doing two git calls, we should either update the comment or instead use those two git calls rather than the slightly less obvious calls we are now doing.

Reply:
To avoid broken HEAD locally, we should checkout master, or the default branch defined in repos.yaml, before git fetch --prune.
I will also improve the comments.
However, if this solution is too scary, we are not obliged to implement it.

Reply:
I think this might work.  It doesn't get rid of the `-u` but it both eliminates the broken `HEAD` state and gets back to a single "over the network" operation.
```
git checkout --detach
git fetch -fup origin +refs/heads/*:refs/heads/* refs/tags/*:refs/tags/*
```

Reply:
Actually, I think it might be workable without the `-u` too.

Reply:
Ah, so the trick is to ensure you are not on a branch before you run the prune?

Reply:
I will try removing the -u.
However, being on master (or default branch) does not resolve the fundamental problem, but should avoid broken HEAD in the local repo.

Reply:
That's the idea -- if it is indeed viable.  It also avoids needing to know the name of other valid branches (which might need to be pruned as well).

Reply:
That what I sayd in my first comment yesterday on DM-3236.
However, removing the -u option gives an error:
`GComoretto@GComorettoMac ~/lsst/pex_exceptions (master) $git fetch -fp origin +refs/heads/*:refs/heads/*  refs/tags/*:refs/tags/*
fatal: Refusing to fetch into current branch refs/heads/master of non-bare repository`

Reply:
It works for me in testing.

```
#!/bin/bash

set -xe

rm -rf gittest
mkdir gittest
cd gittest/
# simulate a canonical remote repo
(
mkdir canonical_repo
cd canonical_repo
git init --bare
)
# add a couple of branches to said repo
(
git clone canonical_repo user_clone
cd user_clone
touch a_file
git add a_file
git commit -m"a commit"
git push origin master
git checkout -b foo
touch anotherfile
git add anotherfile
git commit -m"another commit"
git push origin foo
)
# lsst-build like checkout of foo branch
(
git clone canonical_repo build_clone -b foo
cd build_clone
git branch -a
)
# rm foo branch from canoncial_repo
(
cd user_clone
git push origin :foo
)
# lsst-build like fetch -p
(
cd build_clone
git checkout --detach
git fetch -fp origin +refs/heads/*:refs/heads/* refs/tags/*:refs/tags/*
#git fetch -fup origin +refs/heads/*:refs/heads/* refs/tags/*:refs/tags/*
git branch -a
)
```

Reply:
Even if you do the `git checkout --detach` that @jhoblitt mentions?

Reply:
I tried it locally... from command line on the example repo... and I tried it in the script.

Instead of detach I was thinking to get the default branch using 
```
yaml = self._repos_yaml_lookup(product)
git.checkout(yaml.ref)
```

Reply:
Are you saying that if you do this then you can still do it with a single `git fetch`? Or are you saying that you have to do this in addition to the two fetches? @jhoblitt implies that the detached checkout with one fetch is sufficient.  (the code is already going to checkout the right ref later).

Reply:
@gcomoretto Why does your $PS1 have the string `(master)` in it when the `HEAD` is detached?

Reply:
That was my command line test, I checkout master before running the git fetch -fp....

Reply:
Yes, but what I asked was whether it *works* when you detach rather than checking out master.

Reply:
It doesn't from command line. The second prune is also required.

Reply:
But yes, the option -u is not required anymore.

Reply:
It is working for me with the complete script I posted using git:
```
$ git --version
git version 2.20.1
```
What version of git are you testing with?

Reply:
I think it is better to checkout the default branch instead of detach the HEAD. Is not eups package version id been based to what it is found in the local repository?

Reply:
The aim is to get to a point where a single fetch fixes the problem. Detaching head or not has not bearing on what happens later when lsst-build works out which branch to check out for real.

Reply:
I try local changes in lsst_build, with a checkout --detach, and a single fetch without -u and the main problem (deleted branch been checkout) is not solved.

Reply:
If you add a git checkout foo at the end of you script, you see that the branch is still checkout.
Here the last part of the console output:

```
....
From /Users/GComoretto/temp/test3236-S/gittest/canonical_repo
 - [deleted]         (none)     -> foo
 * [new branch]      master     -> master
+ git branch -a
* (HEAD detached at ffa94c6)
  master
  remotes/origin/HEAD -> origin/master
  remotes/origin/foo
  remotes/origin/master
+ git checkout foo
Branch 'foo' set up to track remote branch 'foo' from 'origin'.
Switched to a new branch 'foo'
Your branch is up to date with 'origin/foo'.
GComoretto@GComorettoMac ~/temp/test3236-S $
```

Reply:
in the same way, when lsst_build is going to prune the local copy, using only one fetch, it will still be able to checkout the branch afterwards.

Reply:
Following Tim's suggestion, I updated your script as follows:

```
...
# lsst-build like fetch -p
(
cd build_clone
git checkout --detach
git fetch --force --prune
git fetch --force --tags
#git fetch -fp origin +refs/heads/*:refs/heads/* refs/tags/*:refs/tags/*
#git fetch -fup origin +refs/heads/*:refs/heads/* refs/tags/*:refs/tags/*
git branch -a
git checkout foo
)
```

and the result is always the same:

```
...
+ git fetch --force --prune
From /Users/GComoretto/temp/test3236-S/gittest/canonical_repo
 - [deleted]         (none)     -> origin/foo
+ git fetch --force --tags
+ git branch -a
* (HEAD detached at 9388277)
  foo
  remotes/origin/HEAD -> origin/master
  remotes/origin/master
+ git checkout foo
Switched to branch 'foo'
Your branch is based on 'origin/foo', but the upstream is gone.
  (use "git branch --unset-upstream" to fixup)
```

the branch is found and checked-out at the end, despite it has been deleted.

Reply:
Tested with git `2.18.0` (the current `lsstsw` pinned version).

### `git_detach.patch`
```diff
diff --git a/python/lsst/ci/prepare.py b/python/lsst/ci/prepare.py
index b559e2b..80fab4f 100644
--- a/python/lsst/ci/prepare.py
+++ b/python/lsst/ci/prepare.py
@@ -383,8 +383,11 @@ class ProductFetcher(object):
             #     git.fetch("origin", "--force", "--prune")
             #     git.fetch("origin", "--force", "--tags")
             # but avoids the overhead of two (possibly remote) git calls.
-            git.fetch("-fup", "origin", "+refs/heads/*:refs/heads/*", "refs/tags/*:refs/tags/*")
-            git.fetch("-p")
+            git.checkout("--detach")
+            git.fetch("-fp", "origin",
+                      "+refs/heads/*:refs/heads/*",
+                      "+refs/heads/*:refs/remotes/origin/*",
+                      "refs/tags/*:refs/tags/*")
 
         # find a ref that matches, checkout it
         for ref in self._ref_candidates(product):
```

### `Dockerfile`
```Dockerfile
cat Dockerfile 
FROM lsstsqre/newinstall

ARG LSST_BUILD_URL=https://github.com/lsst/lsst_build
ARG LSST_BUILD_REF=tickets/DM-3236

USER root

# install modern git needed by pkgautoversion
ARG LSST_GIT_VERSION=2.18.0
ARG GIT_BASE_URL="https://www.kernel.org/pub/software/scm/git"
RUN curl -sSL -O "${GIT_BASE_URL}/git-${LSST_GIT_VERSION}.tar.gz" && \
    tar xzf "git-${LSST_GIT_VERSION}.tar.gz" && \
    cd "git-${LSST_GIT_VERSION}" && \
    ./configure --prefix="/usr/local" && \
    make -j4 && \
    make install

RUN git config --global user.email "you@example.com" && \
    git config --global user.name "Your Name"

# source loadLSST.bash by default for python + eups
RUN echo ". $(pwd)/loadLSST.bash" > "/etc/profile.d/loadLSST.sh"

# install lsst-build
RUN git clone "$LSST_BUILD_URL" -b "$LSST_BUILD_REF"
COPY git_detach.patch lsst_build/git_detach.patch
RUN cd lsst_build && patch -p1 < git_detach.patch
RUN cd lsst_build && git diff
RUN cd lsst_build && python setup.py install

# create bare repo testpk1.git with master & feature1
RUN mkdir -p testpkg1 && \
    cd testpkg1 && \
    git init && \
    mkdir ups && \
    touch ups/testpkg1.table && \
    git add ups/testpkg1.table && \
    git commit -m"fwv" && \
    git checkout -b feature1 && \
    touch feature1 && \
    git add feature1 && \
    git commit -m"add feature1" && \
    git checkout master
RUN mv testpkg1/.git testpkg1.git && \
    rm -rf testpkg1 && \
    cd testpkg1.git && \
    git config --bool core.bare true
RUN echo "testpkg1: $(pwd)/testpkg1.git" > repos.yaml

# create testpkg2 which depends on testpkg1
RUN mkdir -p testpkg2 && \
    cd testpkg2 && \
    git init && \
    mkdir ups && \
    echo "setupRequired(testpkg1)" > ups/testpkg2.table && \
    git add ups/testpkg2.table && \
    git commit -m"fwv" && \
    git checkout -b feature1 && \
    touch feature1 && \
    git add feature1 && \
    git commit -m"add feature1" && \
    git checkout master
RUN mv testpkg2/.git testpkg2.git && \
    rm -rf testpkg2 && \
    cd testpkg2.git && \
    git config --bool core.bare true
RUN echo "testpkg2: $(pwd)/testpkg2.git" >> repos.yaml

# lsst-build "build" dir
RUN mkdir build

# prepare feature1 branch
RUN lsst-build prepare --repos="$(pwd)/repos.yaml" "$(pwd)/build" testpkg2 \
    --ref feature1

# show clone branch state
RUN cd build/testpkg1 && \
    git branch -a
RUN cd build/testpkg2 && \
    git branch -a

# rm feature1 branch from testpkg1
RUN git clone testpkg1.git testpkg1 && \
    cd testpkg1 && \
    git push origin :feature1 && \
    rm -rf testpkg1

# prepare feature1 branch
RUN lsst-build prepare --repos="$(pwd)/repos.yaml" "$(pwd)/build" testpkg2 \
    --ref feature1

# show clone branch state
RUN cd build/testpkg1 && \
    git branch -a
RUN cd build/testpkg2 && \
    git branch -a
```
### `docker build . --no-cache`
```bash
...
Step 22/28 : RUN lsst-build prepare --repos="$(pwd)/repos.yaml" "$(pwd)/build" testpkg2     --ref feature1
 ---> Running in 51794fa8d0f5
            testpkg2:  ok (0.1 sec).
            testpkg1:  ok (0.1 sec).
Removing intermediate container 51794fa8d0f5
 ---> 78d83a8297a1
Step 23/28 : RUN cd build/testpkg1 &&     git branch -a
 ---> Running in 0b1d0fdbd14d
* feature1
  master
  remotes/origin/HEAD -> origin/master
  remotes/origin/feature1
  remotes/origin/master
Removing intermediate container 0b1d0fdbd14d
 ---> 39043926c1e3
Step 24/28 : RUN cd build/testpkg2 &&     git branch -a
 ---> Running in d133f2269f93
* feature1
  master
  remotes/origin/HEAD -> origin/master
  remotes/origin/feature1
  remotes/origin/master
Removing intermediate container d133f2269f93
 ---> 92add6d7d161
Step 25/28 : RUN git clone testpkg1.git testpkg1 &&     cd testpkg1 &&     git push origin :feature1 &&     rm -rf testpkg1
 ---> Running in 334937d7f184
Cloning into 'testpkg1'...
done.
To /opt/lsst/software/stack/testpkg1.git
 - [deleted]         feature1
Removing intermediate container 334937d7f184
 ---> 41d789828666
Step 26/28 : RUN lsst-build prepare --repos="$(pwd)/repos.yaml" "$(pwd)/build" testpkg2     --ref feature1
 ---> Running in ab8658227dea
            testpkg2:  ok (0.1 sec).
            testpkg1:  ok (0.0 sec).
Removing intermediate container ab8658227dea
 ---> 22d6a54d2030
Step 27/28 : RUN cd build/testpkg1 &&     git branch -a
 ---> Running in 571e7ad9097a
* master
  remotes/origin/HEAD -> origin/master
  remotes/origin/master
Removing intermediate container 571e7ad9097a
 ---> a8822db7fa91
Step 28/28 : RUN cd build/testpkg2 &&     git branch -a
 ---> Running in 8cd00bb17d61
* feature1
  master
  remotes/origin/HEAD -> origin/master
  remotes/origin/feature1
  remotes/origin/master
Removing intermediate container 8cd00bb17d61
 ---> 29ef45083e6d
Successfully built 29ef45083e6d

```

Reply:
Congratulation Josh. Your change is of course much better than my first attempt.

Do want to update the comments yourself, since you have done the change, it may be quicker, or do you want me to do it?

Reply:
No thanks, please go ahead.

Reply:
Updates done